src/main/java/com/nimbusds/jose/crypto/impl/ECDH1PUCryptoProvider.java

```diff
@@ -313,26 +310,32 @@ protected JWECryptoParts encryptMulti(final JWEHeader header,
     }
 
     protected byte[] decryptMulti(final JWEHeader header,
-                                  final Map<String, SecretKey> sharedSecrets,
+                                  final List<Pair<UnprotectedHeader, SecretKey>> sharedSecrets,
                                   final List<Recipient> recipients,
                                   final Base64URL iv,
                                   final Base64URL cipherText,
                                   final Base64URL authTag) throws JOSEException {
 
         byte[] result = null;
 
-        for (Map.Entry<String, SecretKey> rs : sharedSecrets.entrySet()) {
+        for (Pair<UnprotectedHeader, SecretKey> rs : sharedSecrets) {
+            String kid = rs.getLeft().getKeyID();
             Base64URL encryptedKey = null;
 
+            if (kid == null) {
+                throw new JOSEException("kid should be specified");
```

Comment:
So, we require that `kid` must be set for every recipient during decryption, but we don't require that during encryption, right? If so, it looks a bit inconsistent and can lead to wrong API usage.